test/suite/test_tiered17.py

```diff
@@ -0,0 +1,122 @@
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# test_tiered17.py
+#    Test that opening a file in readonly mode does not create a new object in tier.
+
+from helper import copy_wiredtiger_home
+from helper_tiered import TieredConfigMixin, gen_tiered_storage_sources, get_conn_config
+from wtscenario import make_scenarios
+import os, wiredtiger, wttest
+
+class test_tiered17(TieredConfigMixin, wttest.WiredTigerTestCase):
+    tiered_storage_sources = gen_tiered_storage_sources()
+    saved_conn = ''
+    uri = "table:test_tiered"
+
+    shutdown = [
+        ('clean', dict(clean=True)),
+        ('unclean', dict(clean=False)),
+    ]
+
+    def conn_config(self):
+        if self.is_tiered_scenario():
+            self.saved_conn = get_conn_config(self) + ')'
+        return self.saved_conn
+
+    scenarios = make_scenarios(tiered_storage_sources, shutdown)
+
+    def get_object_files(self):
+        object_files = []
+        for path, _, files in os.walk('./'):
+            for file in files:
+                if file.endswith('.wtobj'):
+                    object_files.append(os.path.join(path, file))
+        return object_files
+
+    def test_open_readonly_conn(self):
+        # Create and populate a table.
+        self.session.create(self.uri, "key_format=S,value_format=S")
+        c = self.session.open_cursor(self.uri)
+        c["a"] = "a"
+        c["b"] = "b"
+
+        # Do a checkpoint and flush operation.
+        self.session.checkpoint()
+        self.session.flush_tier(None)
+
+        # Add more data but don't do a checkpoint or flush in the unclean shutdown scenario.
+        if not self.clean:
+           c["c"] = "c"
+           c["d"] = "d"
+        c.close()
+
+        obj_files_orig = self.get_object_files()
+
+        # Re-open the connection but in readonly mode.
+        conn_params = 'readonly=true,' + self.saved_conn
+        self.reopen_conn(config = conn_params)
+
+        obj_files = self.get_object_files()
+
+        # Check that no additional object files have been created.
+        self.assertTrue(sorted(obj_files_orig) == sorted(obj_files))
+
+    def test_open_readonly_cursor(self):
+        # Create and populate a table.
+        self.session.create(self.uri, "key_format=S,value_format=S")
+        c = self.session.open_cursor(self.uri)
+        c["a"] = "a"
+        c["b"] = "b"
+
+        # Do a checkpoint and flush operation.
+        self.session.checkpoint()
+        self.session.flush_tier(None)
+
+        # Add more data but don't do a checkpoint or flush in the unclean shutdown scenario.
+        if not self.clean:
+           c["c"] = "c"
+           c["d"] = "d"
+        c.close()
```

Comment:
We can extract the code from the beginning of the method up to this point into a separate method. We could call it `populate`, and call it from both test methods.